modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1127 @@
+// all inital cracks are defined at t_start = t_end = 0
+MeshCut3DUserObject::MeshCut3DUserObject(const InputParameters & parameters)
+  : GeometricCutUserObject(parameters),
+    _mesh(_subproblem.mesh()),
+    _n_step_growth(getParam<unsigned int>("n_step_growth")),
+    _func_x(getFunction("function_x")),
+    _func_y(getFunction("function_y")),
+    _func_z(getFunction("function_z"))
+{
+  _grow = (_n_step_growth == 0 ? 0 : 1);
+
+  if (_grow)
+  {
+    if (!isParamValid("size_control"))
+      mooseError("Crack growth needs size control");
+
+    _size_control = getParam<Real>("size_control");
+  }
+
+  // only the xda type is currently supported
+  MeshFileName xfem_cut_mesh_file = getParam<MeshFileName>("mesh_file");
+  _cut_mesh = libmesh_make_unique<ReplicatedMesh>(_communicator);
+  _cut_mesh->read(xfem_cut_mesh_file);
+
+  // test element type; only tri3 elements are allowed
+  auto elem_it = _cut_mesh->elements_begin();
+  auto elem_end = _cut_mesh->elements_end();
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+    if (cut_elem->n_nodes() != _cut_elem_nnode)
+      mooseError("The input cut mesh should include tri elements only!");
+    if (cut_elem->dim() != _cut_elem_dim)
+      mooseError("The input cut mesh should have 2D elements only!");
+  }
+}
+
+bool
+MeshCut3DUserObject::active(Real time) const
+{
+  return time >= 0;
+}
+
+void
+MeshCut3DUserObject::initialize()
+{
+  // this is a test of all methods relevant to front growth
+  if (_grow)
+  {
+    _stop = 0;
+
+    findBoundaryNodes();
+    findBoundaryEdges();
+    sortBoundaryNodes();
+    refineBoundary();
+
+    for (unsigned int i = 0; i < _n_step_growth; ++i)
+    {
+      findActiveBoundaryNodes();
+
+      if (_stop != 1)
+      {
+        findActiveBoundaryDirection();
+        growFront();
+        sortFrontNodes();
+
+        if (_inactive_boundary_pos.size() != 0)
+          findFrontIntersection();
+
+        refineFront();
+        triangulation();
+        joinBoundary();
+      }
+    }
+  }
+}
+
+bool
+MeshCut3DUserObject::cutElementByGeometry(const Elem * /*elem*/,
+                                          std::vector<CutEdge> & /*cut_edges*/,
+                                          std::vector<CutNode> & /*cut_nodes*/,
+                                          Real /*time*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::cutElementByGeometry(const Elem * elem,
+                                          std::vector<CutFace> & cut_faces,
+                                          Real /*time*/) const
+// With the crack defined by a planar mesh, this method cuts a solid element by all elements in the
+// planar mesh
+// TODO: Time evolving cuts not yet supported in 3D (hence the lack of use of the time variable)
+{
+  bool elem_cut = false;
+
+  const auto elem_begin = _cut_mesh->elements_begin();
```

Comment:
It's important to note here that `_cut_mesh` is not the main FE mesh.  It's a topologically 2D mesh used to define cutting planes for a 3D FE mesh. It will always be significantly smaller than the main solution mesh, and I doubt that it would ever make sense for it to be distributed.